example-plugins/inj-tyfam-plugin/src/Plugin/InjTyFam.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Plugin.InjTyFam (
+  plugin,
+  tcPlugin,
+  ) where
+
+import           Control.Monad (guard)
+import           Data.Maybe (fromMaybe)
+
+import qualified Coercion
+import qualified CoreSyn
+import qualified Plugins
+import           TcEvidence (EvTerm)
+import qualified TcPluginM
+import qualified TcRnMonad
+import qualified TcRnTypes
+import           TcRnTypes (Ct, Xi)
+import           TcType (TcType)
+import qualified TyCoRep
+import           TyCoRep (Coercion)
+import qualified TyCon
+import           TyCon (TyCon)
+import qualified Type
+import qualified Unify
+import           VarEnv (VarEnv)
+import qualified VarEnv
+
+-- | Just 'tcPlugin', with 'Plugins.purePlugin' set
+
+plugin :: Plugins.Plugin
+plugin = Plugins.defaultPlugin
+    {
+      Plugins.pluginRecompile = Plugins.purePlugin
+    ,
+      Plugins.tcPlugin = \_opts -> pure tcPlugin
+    }
+
+-- | A typechecker plugin that decomposes Given equalities headed by
+-- an injective type family,
+-- <https://downloads.haskell.org/ghc/8.6.5/docs/html/users_guide/glasgow_exts.html#injective-type-families>
+--
+-- The plugin user is " opting in " to
+-- <https://gitlab.haskell.org/ghc/ghc/issues/10833>
+
+tcPlugin :: TcRnTypes.TcPlugin
+tcPlugin = TcRnTypes.TcPlugin
+    {
+      TcRnTypes.tcPluginInit = pure ()
+    ,
+      TcRnTypes.tcPluginSolve = \() gs ds ws -> do
+        let
+          fskEnv :: FskEnv
+          fskEnv = mkFskEnv gs
+
+        if null ds && null ws
+          then mapMR (simplifyG fskEnv) gs   -- we handle G-mode
+          else skip   -- GHC already handles W-mode
+    ,
+      TcRnTypes.tcPluginStop = \() -> pure ()
+    }
+
+-----
+
+type M = TcPluginM.TcPluginM
```

Comment:
Maybe stick with the original types for clarity?

Reply:
Done, thanks.